File: node/nodejs-boilerplate/controllers/auth.controller.js
const csrf = require('csrf-token')
const login = require('../models/login');
const user = require('../models/user');
const sanitizer = require('../utils/sanitizer')


/**
 * Login allows the user to authenticate with credentials 
 * and get a token to use on secured endpoints. 
 */
module.exports.login = async(req, res) => {

  const db = req.db
  try{
    const username = sanitizer.checkStringSanity(req.body.username) ? req.body.username : null
    const pwd = sanitizer.checkStringSanity(req.body.password) ? req.body.password : null

    if(username != null && pwd != null){
      
      const query = `SELECT id, password from users where username="${username}";`
      const queryResult = await db.get(query)
      password = (queryResult ||{}).password
      
      if(password==undefined) throw new Error("Please check your credentials and try again")      
      if(password !== pwd) throw new Error("The combination for the given user/password is incorrect")
      

      //create an expiration date-time for the current session (5 minutes ahead of now)
      let now =new Date(Date.now())
      let expiryDateTime = new Date(now.setSeconds( now.getSeconds() + 300 ));
      
      //token creation based on username and expire datetime for the session
      const token = csrf.createSync(`${username}-${expiryDateTime.toString()}`)

      //create session in db and return to token to consumer
      const insertQuery = `INSERT INTO sessions(user_id, token, expire_datetime) VALUES(${queryResult.id},"${token}", "${expiryDateTime}" )`
      const loginResult = await db.run(insertQuery)

      user.token = token
      user.id = queryResult.id
      res.status(200).json(user);


    }
    else{
      throw new Error("Invalid username or Password")   
    }

  }
  catch(err){
    res.status(400).json({"result": err.message})
  }

};

File: java/java-boilerplate/src/main/java/com/asapp/backend/challenge/controller/AuthController.java
package com.asapp.backend.challenge.controller;

import com.asapp.backend.challenge.resources.LoginResource;
import com.asapp.backend.challenge.resources.UserResource;
import com.asapp.backend.challenge.utils.DatabaseConnection;
import com.asapp.backend.challenge.utils.JSONUtil;
import com.asapp.backend.challenge.utils.TokenGenerator;
import com.google.gson.Gson;
import com.google.gson.JsonObject;

import spark.*;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.sql.Timestamp;
import java.util.Calendar;



public class AuthController {

    public static Route login = (Request req, Response resp) -> {
    	LoginResource response =new LoginResource();
    	
    	DatabaseConnection con = DatabaseConnection.getInstance();
    	UserResource user = new UserResource();
		JsonObject jsonReq = new Gson().fromJson(req.body(), JsonObject.class);

		String username = jsonReq.get("username").getAsString();
		String password = jsonReq.get("password").getAsString();
        try{  
    		String q = "SELECT id, password from users where username=\"" + username + "\";";

    		Statement pstmt = con.getConnection().createStatement();  
            ResultSet rs = pstmt.executeQuery(q);  
            
            //user not found
    		if (!rs.next()) {
    			JsonObject jsonErrUserNotFound = new JsonObject();
    			jsonErrUserNotFound.addProperty("result", "Please check your credentials and try again.");
    			JsonObject jsonErrUserNotFoundResult = new Gson().fromJson(jsonErrUserNotFound, JsonObject.class);
    			Spark.halt(401, jsonErrUserNotFoundResult.toString());
    		}
    		
    		//user/pass combination wrong
    		String pass = rs.getString("password");
    		Long id = rs.getLong("id");
    		
    		if(!(pass.equals(password))) {
    			JsonObject jsonErrUserPassCombination = new JsonObject();
    			jsonErrUserPassCombination.addProperty("result", "The combination for the given user/password is incorrect.");
    			JsonObject jsonErrUserPassCombinationResult = new Gson().fromJson(jsonErrUserPassCombination, JsonObject.class);
    			Spark.halt(401, jsonErrUserPassCombinationResult.toString());    			
    		}
    		
    		
    		//all ok, create an expiration timestamp 5 minutes (300 seconds) ahead of now 
    	    Calendar cal = Calendar.getInstance();
    	    cal.add(Calendar.SECOND, 300);
    	    Timestamp timestamp = new Timestamp(cal.getTime().getTime());
    		
    		
    	    //create a token
    	    String token = TokenGenerator.generateNewToken();
    	    
    	    //insert session into database
    		String iq = "INSERT INTO sessions(user_id, token, expire_datetime) VALUES(" + id +",\"" + token +"\", \"" + timestamp+"\" )";
    		PreparedStatement insertStatement = con.getConnection().prepareStatement(iq);  
            int res = (insertStatement.executeUpdate());
        
            
            
            //read id of the inserted record
            ResultSet generatedKeys = insertStatement.getGeneratedKeys();
            generatedKeys.next();
            
            //sets the login model with the result and send the response
            response.setId(generatedKeys.getLong(1));
            response.setToken(token);
            
		
    		
    		
        } catch (SQLException e) {  
        	JsonObject err = new JsonObject();
        	err.addProperty("Result", "Error");
        	JsonObject jsonObjectResult = new Gson().fromJson(err, JsonObject.class);
      		Spark.halt(401, jsonObjectResult.toString());
            System.out.println(e.getMessage());  
        }  
        return JSONUtil.dataToJson(response);   
    };

}
